### PR TITLE
Extend ChannelType by command options

### PR DESCRIPTION
**Pull request by htreu**
This addresses https://github.com/eclipse/smarthome/issues/5099 by adding a command description with command options along with the state description.
Command options will give a hint to UIs about the specific commands a channel provides. Command options could be rendered as a drop down and also represent the current state or rendered as push buttons to simply send a command to the ThingHandler.
The infrstructure basically copies the StateDescription infrastructure with CommandDescriptionProviders and an `DynamicCommandDescriptionProvider` interface for bindings to hook in and provide dynamic command options.

Signed-off-by: Henning Treu <henning.treu@googlemail.com>

### CODE REVIEW

**Comment by maggu2810** [comment]
I have a PR for the thing.xml test migration in the queue.
Just need to wait until #535 has been merged.

Please integrate your test changed in the migrated ones.

**Comment by martinvw** [comment]
@htreu could you please rebase? and apply the changes as requested by @maggu2810

**Comment by htreu** [comment]
ping @maggu2810 & @martinvw :-)

**Comment by martinvw** [comment]
@openhab/core-maintainers

File: bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ChannelCommandDescriptionProvider.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides the {@link ChannelType} specific {@link CommandDescription} for the given item name and locale.
+ *
+ * @author Henning Treu - Initial contribution
+ *
+ */
+@NonNullByDefault
+@Component(immediate = true)
```
Comment (maggu2810) [inline bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ChannelCommandDescriptionProvider.java]: Why `immediate`? I assume it makes only sense if someone consumes that provider.

**Comment by maggu2810** [comment]
> ... or rendered as push buttons to simply send a command to the ThingHandler.

Do I get you right, that this is just a rendering option for the UI?
If my reading has been correct, the channel itself still still provides a state (`ChannelKind.STATE`) and is not stateless. Correct?

**Comment by htreu** [comment]
> Do I get you right, that this is just a rendering option for the UI?

Yes, this is correct. The state is still provided and it's up to the UI how to deal with it.

File: bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ChannelCommandDescriptionProvider.java (hunk 1 above)
Comment (htreu) [reply bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ChannelCommandDescriptionProvider.java]: correct, I fixed this with the latest commit.

**Comment by cweitkamp** [comment]
@kaikreuzer What has to be done to deploy the changed XSD file to the website (https://www.openhab.org/schemas/thing-description-1.0.0.xsd)? How can I help?

Do we have to change the namespace in those files to: "https://openhab.org/schemas/XXX/v1.0.0"?

**Comment by kaikreuzer** [comment]
@cweitkamp See https://github.com/openhab/website/pull/157

**Comment by cweitkamp** [comment]
@kaikreuzer Thanks. Now it is working in the IDE.

**Pull request merged**